text/0158-dynamic-span-creation.md

```diff
@@ -0,0 +1,109 @@
+Instead, OpenTelemetry simply forwards tracing headers and performs other activities not related to span creation, such as injecting span contexts into logging libraries.
+During crisis situations, operators can dynamically activate the OpenTelemetry instrumentation to acquire additional detail on problematic services, but only for short periods of time.
+As Castanheira puts it:
+
+> Despite the increasing number of uses for distributed tracing, it remains, in the wild, focused on the more common and general use cases (such as latency profiling for troubleshooting slow requests).
+
+For this reason, he argues, applications are typically instrumented very coarsely, and perhaps only for a few central workflows. However, he proceeds:
+
+> During crises, operators may be willing to accept a decrease in performance in exchange for a larger wealth of information, but cannot act on this intent because of the static tracing instrumentation.
+
+## Explanation
+
+In order for operators to configure the expressiveness of tracing at runtime, they need a way to communicate with the instrumented applications.
+As inspired by Jaeger, this could be implemented by allowing for OpenTelemetry Collectors to inform applications about the desired level of detail, perhaps using polling. We believe that the level of detail (LoD) should be configurable at at least three levels:
+
+- `PROPAGATION_ONLY`: Only propoagate headers, and suppress the creation of all spans. This will work well with service meshes.
+- `EXTERNAL_SPANS`: Propagate headers, and generate additional client, server, consumer, and producer spans. This will enhance the traces generated by service meshes, but only when the performance impact can be justified.
+- `ALL_SPANS`: Propagate headers, and generate spans at any level of detail
+
+We can imagine the following uses for the three LoDs:
+
+| `PROPAGATION_ONLY`          | `EXTERNAL_SPANS`           | `ALL_SPANS`                  |
+| --------------------------- | -------------------------- | ---------------------------- |
+| Envoy generates server span |                            |                              |
+| OTel span suppressed        | OTel generates server span | OTel generates server span   |
+| OTel span suppressed        | OTel span suppressed       | OTel generates internal span |
+| OTel span suppressed        | OTel generates client span | OTel generates client span   |
+| Envoy generates client span |                            |                              |
+
+## Internal details
+
+This proposal would entail adding a LoD configuration that can be passed to applications via the collector.
+During span creation, the instrumentation library would have to consider the current LoD, and decide whether to suppress the creation of spans, similarly to how it may decide whether a span should be sampled.
```

Comment:
I think in practice the desired verbosity level would need to be propagated in the trace context.

Reply:
I understand why it would be of interest to use the existing capability to transmit state in the trace context. Initially, the thought was to configure instrumentation verbosity on a service-by-service level, in which case I do not believe that the trace context is sufficient. However, we are also open to the idea of configuring trace verbosity per workflow, if that is more in line with the design of OpenTelemetry. In fact, workflow-based trace verbosity may even be the better option, and reduce the performance impact during debugging periods! We will consider this and update the proposal soon.